pkg/gomod/modules_test.go

```diff
@@ -0,0 +1,192 @@
+				}
+			}
+		})
+	}
+}
+
+func TestModules(t *testing.T) {
+	tests := map[string]struct {
+		files    []string
+		domain   string
+		wantPkgs map[string][]string
+		wantDeps []string
+		wantErr  bool
+	}{
+		"example1, example2, knative.dev": {
+			files:  []string{"testdata/gomod.example1", "testdata/gomod.example2"},
+			domain: "knative.dev",
+			wantPkgs: map[string][]string{
+				"knative.dev/test-demo1": {"knative.dev/eventing", "knative.dev/pkg", "knative.dev/serving", "knative.dev/test-infra"},
+				"knative.dev/test-demo2": {"knative.dev/discovery", "knative.dev/pkg", "knative.dev/test-infra"},
+			},
+			wantDeps: []string{"knative.dev/eventing", "knative.dev/pkg", "knative.dev/serving", "knative.dev/test-infra", "knative.dev/discovery"},
+		},
+		"example1, example2, k8s.io": {
+			files:  []string{"testdata/gomod.example1", "testdata/gomod.example2"},
+			domain: "k8s.io",
+			wantPkgs: map[string][]string{
+				"knative.dev/test-demo1": {"k8s.io/api", "k8s.io/apimachinery", "k8s.io/client-go"},
+				"knative.dev/test-demo2": {"k8s.io/api", "k8s.io/apimachinery", "k8s.io/client-go"},
+			},
+			wantDeps: []string{"k8s.io/api", "k8s.io/apimachinery", "k8s.io/client-go"},
+		},
+		"dup example1, knative.dev": {
+			files:  []string{"testdata/gomod.example1", "testdata/gomod.example1"},
+			domain: "knative.dev",
+			wantPkgs: map[string][]string{
+				"knative.dev/test-demo1": {"knative.dev/eventing", "knative.dev/pkg", "knative.dev/serving", "knative.dev/test-infra"},
+			},
+			wantDeps: []string{"knative.dev/eventing", "knative.dev/pkg", "knative.dev/serving", "knative.dev/test-infra"},
+		},
+		"bad example": {
+			files:   []string{"testdata/gomod.example1", "testdata/bad.example"},
+			domain:  "knative.dev",
+			wantErr: true,
+		},
+		"no domain": {
+			files:   []string{"testdata/gomod.example1", "testdata/gomod.example2"},
+			wantErr: true,
+		},
+		"missing file": {
+			files:   []string{"testdata/gomod.example1", "does-not-exist"},
+			domain:  "knative.dev",
+			wantErr: true,
+		},
+		"no file": {
+			domain:  "knative.dev",
+			wantErr: true,
+		},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			pkgs, deps, err := Modules(tt.files, tt.domain)
+			if (tt.wantErr && err == nil) || (!tt.wantErr && err != nil) {
+				t.Errorf("unexpected error state, want error == %t, got %v", tt.wantErr, err)
+				return
+			}
+			if len(pkgs) != len(tt.wantPkgs) {
+				t.Errorf("Modules() []pkgs length mismatch; got %v, want: %v", pkgs, tt.wantPkgs)
+				return
+			}
+			for name, gots := range pkgs {
+				if want, found := tt.wantPkgs[name]; !found {
+				} else if len(gots) != len(want) {
+					t.Errorf("unexpected pkgs[%q].dep length mismatch; got %v, want: %v", name, gots, want)
+					continue
+				} else {
+					for i, got := range gots {
+						if want := tt.wantPkgs[name][i]; got != want {
+							t.Errorf("unexpected pkgs[%q].dep[%d]; got %q, want: %q", name, i, got, want)
+						}
+					}
```

Comment:
do you want to just sort and cmp.Equal? There's also a way where you can write a custom comparator to sort on the fly in the cmp.

Reply:
oh! yes. thanks.